ext/RastersArchGDALExt/cellarea.jl

```diff
@@ -0,0 +1,98 @@
+        beta2 = _great_circle_bearing(GI.x(point), GI.y(point), GI.x(nextpoint), GI.y(nextpoint)) 
+        angle = acos(cos(-beta1)*cos(-beta2) + sin(-beta1)*sin(-beta2))
+        area += angle
+
+        prevpoint = point
+        point = nextpoint
+    end
+    
+    return area*radius^2
+end
+
+_area_from_coords(transform, geom; radius) = _area_from_coords(transform, GI.trait(geom), geom; radius)
+function _area_from_coords(transform::ArchGDAL.CoordTransform, ::GI.LinearRingTrait, ring; radius)
+    points = map(GI.getpoint(ring)) do p 
+        t = ArchGDAL.transform!(ArchGDAL.createpoint(p...), transform)
+        (deg2rad(GI.x(t)), deg2rad(GI.y(t)))
+    end
+    return _area_from_rads(GI.LinearRing(points); radius)
+end
+
+# For lat-lon projections. Get the area of each latitudinal band, then multiply by the width
+function _area_from_lonlat(lon::XDim, lat::YDim; radius)
+    two_pi_R2 = 2 * pi * radius * radius
+    band_area = broadcast(DD.intervalbounds(lat)) do yb
+        two_pi_R2 * (sin(deg2rad(yb[2])) - sin(deg2rad(yb[1])))
+    end
+    
+    broadcast(DD.intervalbounds(lon), band_area') do xb, ba
+        abs(xb[2] - xb[1]) / 360 * ba
+    end
+end
+
+function cellarea(dims::Tuple{<:XDim, <:YDim}; radius = 6371008.8)
+    # check the dimensions 
+    isnothing(crs(dims)) && _no_crs_error()
+    any(d -> d isa Points, sampling.(dims)) && throw(ArgumentError("Cannot calculate cell size for a `Raster` with `Points` sampling."))
+
+    areas = if _isdegrees(crs(dims)) # check if need to reproject
+        _area_from_lonlat(dims...; radius)
+    elseif !isnothing(mappedcrs(dims)) && _isdegrees(mappedcrs(dims))
+        _area_from_lonlat(reproject(dims; crs = mappedcrs(dims))...; radius)
+    else
+        xbnds, ybnds = DD.intervalbounds(dims)
+        ArchGDAL.crs2transform(crs(dims), EPSG(4326), order = :trad) do transform
+            [_area_from_coords(
+                transform,         
+                GI.LinearRing([
+                    (xb[1], yb[1]), 
+                    (xb[2], yb[1]), 
+                    (xb[2], yb[2]), 
+                    (xb[1], yb[2]),
+                    (xb[1], yb[1])
+                ]); 
+                radius
+                )
+                for xb in xbnds, yb in ybnds]
+        end
+    end
+
+    return Raster(areas, dims)
+end
+function cellarea(x::Union{<:AbstractRaster, <:AbstractRasterStack, <:RA.DimTuple}; kw...)
+    cellarea(dims(x, (XDim, YDim)); kw...)
+end
+
+# TODO: put these in ArchGDAL
+_isgeographic(crs) = _isgeographic(ArchGDAL.importCRS(crs))
+_isgeographic(crs::AG.ISpatialRef) = AG.GDAL.osrisgeographic(crs) |> Bool
+
+_isdegrees(crs) = _isdegrees(ArchGDAL.importCRS(crs))
+function _isdegrees(crs::AG.ISpatialRef)
+    _isgeographic(crs) || return false
+    pointer = Ref{Cstring}()
+    result = AG.GDAL.osrgetangularunits(crs, pointer)
+    return unsafe_string(pointer[]) == "degree"
+end
```

Comment:
Nice!

Reply:
```

julia> _isdegrees(GFT.EPSG(3857))
false

julia> _isgeographic(GFT.EPSG(3857))
false

```

The code here is web mercator, which is known remappable to long/lat.  So we might need another check...

Reply:
But it's still in meters so we don't want to use the _area_from_lonlat for that projection. I tried for EPSG(4326) and EPSG(4258) which both returned true. It's the best Raf and I could come up with and I think it works, but the documentation on these functions is pretty sparse and there are just so many cases.

Reply:
https://github.com/rafaqz/Rasters.jl/blob/8d13c1887f90444d24e3c19ecbff42c8d5f0de68/ext/RastersArchGDALExt/reproject.jl#L7-L9

is the approximate check that's currently being used I think, we could make it a bit more robust

Reply:
Maybe we should keep this for a separate PR, since it's a bigger discussion. If we've missed any cases where we're not using the shortcut for lon/lat projections, then cellarea will use the fallback which is still really accurate (except for very small gridcells)

Reply:
If the fallback is what we previously had then I would say we need a better solution. It's pretty common for geospatial datasets to have pixel spacings of < 10 m.. but maybe I misunderstood.

Reply:
The fallback is still to use Girard's theorem, which is what we had previously. It might be slightly more accurate now that we use sind and cosd, but for such small pixel spacings my guess is that it will still suffer from numerical stability issues. But I don't really know what to use instead

Reply:
> If the fallback is what we previously had then I would say we need a better solution. 

What is you idea for this @alex-s-gardner ?

Reply:
The previous precision errors were pretty strange if everything was being maintained at Float64. Maybe the issue was the sind and cosd and then it's a non-issue. If not then I guess we fallback to a simple width * height for very small gidcells? Though that seems like a hack.

Reply:
Using Girard's theorem means we sum the angles of the gridcell and see how much bigger that is than it would be on a plane. For a 100m gridcell that is in the difference in the angles is in the order of magnitude e-12, and we're doing lots of sin and cos to get there, so I do think it could be a numerical accuracy issue. 

I don't think width * height will be so easy, though. After reprojecting the gridcells won't be rectangular anymore, so we need to calculate a straight angle. And on this scale I don't know if numerical accuracy will then again become an issue?

Reply:
Is there a crossover point where the angles are close enough to 90 degrees that we can switch to just doing h * w?

Reply:
Not in any predictable way, right? The size of the gridcell doesn't change the angles.